### ISSUE TITLE
Pay back technical debt for 3.2.0 (EDIT: 3.3.0)

### ISSUE THREAD

**Issue by Blaisorblade** [farkam135/GoIV#296]
Development is going pretty fast and some regressions keep coming back.

We discussed this in chat and I suspect we should have a 3.1 release soon with critical features and possibly plan a maintenance branch for any regressions there, and focus 3.2 on
- modularizing (proposed by e.g. @Blaisorblade). EDIT see also #278 to split Pokefly.
- adding tests (proposed by e.g. @Blaisorblade)
- any other clean up, re-evaluate the data structures/representations, make a release and then start features again (proposed by @sarav).
- generally, "paying technical debt"

so that we can later start adding more features again.

I also wonder if we should have a 3.0.2 release branch with stuff like #269.

BTW, if we want up-to-date translations it could also be good to stabilize UI strings, but I'm not sure we can do that yet. What do others think?
Tagging at least @sarav @rhari991 @harkin @nahojjjen @farkam135 @oskomonzon @DJCrapsody, but others are welcome to chime in. Especially if such a plan would help them to contribute—but even users' opinion can be relevant since they're impacted.

EDIT: here's a list of more specific proposals, by @sarav:
- [ ] Move pokedex and pokemap into Pokemon (sure, rename it PokemonSpecies if you really want to)
- [ ] Remove all null checks that we think shouldn't happen and let it crash.
- [ ] Have a MAX 1 global variable to denote the dialog boxes are up and the IV button shouldn't show up. Ideally no global variables. We have like 3 of those -- infosent, infoshown, readyforscreenshot, etc.
- [ ] Move UI code out of Pokefly #278 
- [x] FIX screenshot handling code #149 -- it has a lot of code (suggested by me to get it to work) that doesn't work (ironically, on my phone too). @farkam135 -- can you please? You know this code the best.
- [x] Move screenshot handling code out of Pokefly
- [x] Fix issue #183 and clean up the ordering in Pokefly too (hopefully there not much start up code in MainActivity after this clean up)

I'm adding ideas below.

**Comment by sarav** [comment]
I'm not sure we need a separate maintenance branch. Just pause PR till a set of listed technical debt is paid off. I don't expect it'll take way too long. I think letting in bug fixes in that time might be ok. Not a strong opinion, just how I would do it.

Btw, I won't be contributing much for about a week. So, I feel like I'm ducking when I should be helping. Sorry about that.

**Comment by Blaisorblade** [comment]
FWIW I totally expect 3.2 to take at least a week, and the past suggests we might have regressions with 3.1.0. Especially if people rush to get features in because afterwards they can't.

**Comment by rhari991** [comment]
The refefactoring would definitely be helpful, as the MainActivity and PokeFly classes are pretty big now.

**Comment by oskomonzon** [comment]
I agree, we should freeze features for a while and take care of the technical debt.

**Comment by sarav** [comment]
Couple of things I'd like to see (in no particular order) happen as part of the debt repayment:

- [ ] Move pokedex and pokemap into Pokemon (sure, rename it PokemonSpecies if you really want to)
- [ ] Remove all null checks that we think shouldn't happen and let it crash.
- [ ] Have a MAX 1 global variable to denote the dialog boxes are up and the IV button shouldn't show up. Ideally no global variables. We have like 3 of those -- infosent, infoshown, readyforscreenshot, etc.
- [ ] Move UI code out of Pokefly #278 
- [ ] FIX screenshot handling code -- it has a lot of code (suggested by me to get it to work) that doesn't work (ironically, on my phone too). @farkam135 -- can you please? You know this code the best.
- [ ] Move screenshot handling code to Pokefly

**Comment by Blaisorblade** [comment]
Since we agreed on the release plan, renaming.

**Comment by sarav** [comment]
Btw, rest of the devs. please do list anything else you think should be part of refactor/tech debt.

@Blaisorblade, feel free to move my list into the top comment if you want. I don't mind.

**Comment by nahojjjen** [comment]
@sarav I agree with all of your list except 

> Move UI code out of Pokefly #278

I think we should break Pokefly into four classes, something like
PokeflyButton - A class representing the flying IV button
PokeflyInput - A class representing the input dialog that appears after pressing the iv button
PokeflyResults - A class representing the results after user has given input
PokeflyResultsExpandedIVs - A class representing the list of all IV combinations

**Comment by DJCrapsody** [comment]
We still need Pokefly to exist as a service, possibly renaming to PokeflyService to easily identify it as a service class just like how activities are identified with _Activity.

Having new classes to accommodate individual feature will help the project to become modular and each class would be easily identified and managed. With those classes included, they should be utilized only by PokeflyService.

Right now we need to ensure that MainActivity **should not**:
* Handle anything else other than the UI thread in start page
* Handle any tasks in the background (PokeflyService should be doing this)
* Handle nullifying or exitting of any object or features (OCRHelper should fire tesseract.end() when PokeflyService is stopped)

**Comment by sarav** [comment]
I'm not sure you are even disagreeing with me. You are just renaming pokefly to pokeflyservice. The only difference is you saying the IV button should be a separate class. I think I said elsewhere that IV could stay in pokefly **for now**. The reason being I want to find out more about how much overhead it'll have if we split it out into an activity. It's something that appears and disappears often. That is expensive with activities. The rest of the UI doesn't change as often. I think we should finish moving the rest of the UI and then use that learning experience to decide what to do with IV button.

Also, the OCR stuff is already moved out of MainActivity unless I missed something in my refactor.

**Comment by DJCrapsody** [comment]
@sarav no disagreement, just added some points if we ever split Pokefly to four classes as suggested by @nahojjjen. Just stating some reminders for MainActivity.

**Comment by sarav** [comment]
So, how's this been going? Want me to help or has anyone done most of the work already?

**Comment by sarav** [comment]
If someone has done most of the work, I don't mind reviewing it too.

**Comment by nahojjjen** [comment]
We started a branch that didn't get very far. Then we added extra appraisal input and some other stuff, so the refactoring branch got conflicted, and it probably wont be worked on more.

We have however done some pretty big changes in pokefly and some other classes by splitting up very large methods & other tweaks (me and blaisorblade mostly). So while pokefly is way too big, the content inside pokefly has become quite a bit more manageble.

**Comment by oskomonzon** [comment]
I started working in the window management from Pokefly making small changes and separating the layouts, mostly testing at the moment.
I'll probably ask @nahojjjen for help and insight with some stuff. Pokefly's UI control methods all over the place.

**Comment by Blaisorblade** [comment]
@sarav I think that branch https://github.com/farkam135/GoIV/tree/breakApartPokefly was useful to get started—but we won't reuse the actual commits. I confess that without you this hasn't progressed much.

Meanwhile, PR https://github.com/farkam135/GoIV/pull/371 for notification actions is also somewhat related to the intent flow.

**Comment by sarav** [comment]
Why did we add more appraisal features? :(

We can't keep adding more features if we want to finish the clean up.

@oskomonzon can you push your changes as a PR so that I can see where it's going? Maybe I can do a hacky separation of UI code from Pokefly so that people can work on the UI code clean up and Pokefly clean up separately?

@nahojjjen Has the big changes to Pokefly merged? If not, should I wait for it or ignore and do the refactor on my end? Either way, trying to avoid conflicts.

**Comment by sarav** [comment]
@Blaisorblade That branch has some good commits. So, maybe I'll cherry pick the good ones, make my changes and then send a PR.

**Comment by oskomonzon** [comment]
@sarav will do as soon as I have a moment at home to end splitting the layouts into independant files and make some small Pokefly tweaks.

Sorry for slowing down, I'll be out this weekend and I need to make sure everything's on point at work.

**Comment by Blaisorblade** [comment]
> We can't keep adding more features if we want to finish the clean up.

Basically without you nobody was able to do the clean up, and adding appraisals was relatively easy anyway. So the choice was between doing nothing and adding a (IMO well-isolated) feature.

**Comment by sarav** [comment]
@oskomonzon , I was asking if **I** should do a hacky UI split if your UI clean up will take time. I don't think I'll call your UI code changes a "hack" 😃 I think your code would be better than my hacky attempts at it.

@Blaisorblade Referring to the chat in Discord, I can see the need for all the commits, but I'm not sure I agree (yet?) on how some of the changes were done. Since the clean up branch kinda stopped, let me try to make a new fork/branch and pull in the non-conflicting commits, make any tweaks to those commits, add my commits, etc and send a PR. Will probably be a few days (probably the weekend). Doesn't have to be my way -- if you don't like how I had done something, we can throw that stuff out and do it differently.

**Comment by Blaisorblade** [comment]
TL;DR. Can I merge https://github.com/farkam135/GoIV/pull/382/commits/e3e4c86fae8c77afdb68c3f489c00c336202fa91 quickly? That should IMHO be rather uncontroversial. I'll wait for an answer, but most other stuff right now should probably wait (unless we're sure it won't conflict).

@sarav I'm not sure what you're suggesting, or which commits you're talking about. You want to rework `breakApartPokefly`? Go ahead, nobody is merging that.

Or do you object to refactorings in `miscRefactorings`? The only refactoring commit in `miscRefactorings` is https://github.com/farkam135/GoIV/pull/382/commits/e3e4c86fae8c77afdb68c3f489c00c336202fa91; we agreed merging it, then new features were written on top of that (which is surely not ideal) to minimize conflicts.

I'd like to merge that first unless you have objections.

> Eg: moving the UI code into functions could be one commit that'd be part of a series of commits that are moving UI code to another class, etc

It's nicer to do refactorings at once, but commits that move stuff around often need be merged quickly or trigger conflicts. And fixing merge conflicts in commits that move code around... is extremely error-prone.

Anyway, right now we are all too busy to merge stuff—as long as that's the case there's no hurry. But otherwise there's lots to do.

**Comment by sarav** [comment]
That one commit looks fine. There's one function that's practically 2 lines and only used once.  I think that's an overkill, but that's alright. Go ahead and merge that commit. 

I was talking about the miscRecactorings branch. I thought it had more than just that one commit.

**Comment by Blaisorblade** [comment]
Notwithstanding the misnomer, all other commits in that branch are about features. I'd fix it now but in a hurry, sorry.

**Comment by oskomonzon** [comment]
@sarav of course! Feel free to exert whatever hackery you deem necessary. I'll be out at least a couple of days starting from tomorrow afternoon, so don't worry about running into conflicts, I don't think I'm gonna be back until Monday.

**Comment by lennykano** [comment]
> Remove all null checks that we think shouldn't happen and let it crash.

I'd suggest using the `@Nullable`/@NonNull` support annotations on methods/parameters/fields throughout the project to improve nullability handling. Much harder to accidentally forget a null check in the future. Or even better, use `lombok.NonNull` which will automatically generate a `NullPointerException` if something violates the contract, and therefore abort early.

**Comment by sarav** [comment]
@lennykano No clue how to use it. Pointers to docs? I don't generally code in Java or for Android. So, not up to date with the fancy tools.

**Comment by lennykano** [comment]
http://tools.android.com/tech-docs/support-annotations (have a look at the Nullness Annotations section)

It's basically a way to tell the IDE whether a return value, parameter or field can possibly be null or nonnull so that i.e. if you try to dereference a field/parameter/method return value marked `@Nullable` without first checking it is not null, the IDE will flag it with a warning. `@NonNull` is just a way for you to say "this value will never be null", and if you try to pass in null/something marked `@Nullable` the IDE will again flag that it's incorrect.

The only difference with using `lombok.NonNull` instead of the support version is that Lombok uses it's magic to weave in an `if (x == null) { throw new NullPointerException() }` to force the application to abort early if the value is null when it shouldn't be.

Very useful if you start using them from the beginning, so that every single field, method, parameter is annotated, you'll never run into a `NullPointerException`. Implementing it project wide after the fact is probably a lot more work though.

**Comment by sarav** [comment]
@Blaisorblade, looks like I can't edit the checklist. You might want to check the 2 items I checked off on my comment.

**Comment by NightMadness** [comment]
Had a thought:
Maybe as part of unit tests we could have some sample bitmaps scanned from diffrent android devices. if we have a set of screenshots tests we will be more confident in refactoring and new ocr features. This will cost us more repo space, we could probably compress the images.

**Comment by nahojjjen** [comment]
Many of the things brought up here are solved, some are only partially solved. It's difficult to keep track of the progress with an issue this big, so I'm closing it. Parts of this issue exists as other issues, if you feel some are missing, please create them.
